src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationCodeFixProvider.cs

```diff
@@ -253,15 +271,14 @@ private static IEnumerable<SyntaxTrivia> MassageTrivia(IEnumerable<SyntaxTrivia>
         }
 
         private async Task<bool> SemanticsChangedAsync(
-            Document document,
-            VariableDeclarationSyntax declaration,
-            ExpressionSyntax invocationOrCreation,
-            TypeSyntax newType,
-            IdentifierNameSyntax identifier,
-            DeclarationExpressionSyntax declarationExpression,
+            Document document, 
+            List<(ILocalSymbol, DeclarationExpressionSyntax, IdentifierNameSyntax)> declarationsToAdd, 
+            ExpressionSyntax invocationOrCreation, 
             CancellationToken cancellationToken)
```

Comment:
note: overall, i think much of this logic is pretty overwrought.  It seems to be trying to reimplement what we already have with ApplySemanticEditsAsync.  I'd prefer to use that helper as it's more tried and true.

Effectively, all you have to do is get the nodes marked by the original diagnostics.  You then say you want to ApplyMethodBodySemanticEdit.  Then that helper will do the grouping for you, where diagnostics in different methods can all apply independently.  But diagnostics within a method apply one at a time.  Then, all you need to do is validate each applied diagnostic to make sure it's not changing semantics if it uses 'var'.

Reply:
Let me know if you need more info.  I'm also available on gitter.im to discuss things more!

Reply:
It seems to me that `ApplyMethodBodySemanticEdit` is more expensive then the current approach. That method seems to be [forking a new semantic model](([http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/SyntaxEditorExtensions.cs,109](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/SyntaxEditorExtensions.cs,109))) after every edit within the method body, while current approach creates a very lightweight speculative semantic model off the original semantic model to validate semantic changes (unless I read the code incorrectly).

---
In reply to: [230210774](https://github.com/dotnet/roslyn/pull/30917#discussion_r230210774) [](ancestors = 230210774)

Reply:
> That method seems to be forking a new semantic model after every edit within the method body

Yes.  it would be more expensive in the case of multiple edits in the same *statement* boundary (i.e. precisely this case).  But this is the outlier case. The majority of cases would be those that don't involve many edits within the same statement boundary.

Reply:
The SemanticEdit helper straddles a line of allowing code to be very simple (so that all the features that use it don't have to write complex logic for handling potential semantic collisions), while also being fast for the common case of no collisions.

Cost is only paid per collision (which in general is rare), with the benefit of having much simpler individual features that can share this sort of approach instead of having to individually do it themselves.  

Does that make sense?

Reply:
Yes, that sounds reasonable. I am wondering if the helper can have an overload that supports the specific scenario that Andrew encountered here and avoid forking semantic model and provide him a speculative model for every statement edit. Regardless, I agree that taking a simplest approach is suitable here.

---
In reply to: [230226018](https://github.com/dotnet/roslyn/pull/30917#discussion_r230226018) [](ancestors = 230226018)

Reply:
> the specific scenario that Andrew encountered here and avoid forking semantic model and provide him a speculative model for every statement edit. 

I think that would definitely be possible.  IIRC, there was some reason i couldn't use that before. Unfortunately i don't remember what that issue was now :( (sorry!).

We could start by just making it *always* give out a speculative model.  Then see what breaks.  If nothing breaks, then we'll just change to that.   If something breaks, then we make this a knob on the API.  So everyone who can use speculative models can use them for the perf boost.

Reply:
Ok.  IIRC, the issue with speculative models was specifically about when you were making a change that was introducing a new variable in scope (like a declaration-expression).  You would end up with a collision (because the original semantic model had a variable with that name) that your speculated code now interfered with.  And even if the variables had the same name+type, you'd often get error symbols, which would make it seem as if the edit was bad. 

It would be interesting to see if that is still an issue.  And, again, i'm totally supportive of this API using speculative models for perf if the caller knows that's safe for their scenario! :)

Reply:
@ryzngard is there enough info here for you to go off of?

Reply:
Yea, I'm working on something more elegant that I think can drive meaningful conversation. Hopefully get an update today based on some of this feedback